### ISSUE TITLE
loss curve in pretraining BERT is very strange

### ISSUE THREAD

**Issue by zjujh1995** [NVIDIA/Megatron-LM#86]
We can't re-produce the result of pretraining BERT in paper "Megatron-LM", and we've observed a very strange loss curve. The loss converged slowly in the beginning. After several k steps, the loss began to speed up to converge.

As we know, loss usually converges fast in the beginning and slows down gradually during training procedure. Therefore, we're very confused to this special loss curve. May somebody finds the reason?

All the parameters are kept the same as in the shell script "pretrain_bert.sh". Or will the curve be more normal if we use the parameters in the original paper instead of the shell script?

Many thanks!

**Comment by zjujh1995** [comment]
BTW, we've also trained GPT model based on Megatron, and the loss is normal as we expect.

**Comment by rhythmswing** [comment]
The model learns initially with a small learning rate which gradually grows to the specified rate. This initial warm-up phase is to prevent the model from overfitting to early samples and empirically useful when you scale to a very large batch size and learning rate. You can change warm-up params to determine how long the phase is

**Comment by zjujh1995** [comment]
> The model learns initially with a small learning rate which gradually grows to the specified rate. This initial warm-up phase is to prevent the model from overfitting to early samples and empirically useful when you scale to a very large batch size and learning rate. You can change warm-up params to determine how long the phase is

Thanks for your reply. However, we've also done an experiment without LR schedualer, and the loss curve was still strange as I described above. On the other hand, the loss curve of GPT model is smooth and normal all the time (w or w/o LR schedualer). Only the one of BERT is confusing.

We've double-checked all our experimental settings, and there're also other researchers except for us facing the same confusing question.

**Comment by rhythmswing** [comment]
> > The model learns initially with a small learning rate which gradually grows to the specified rate. This initial warm-up phase is to prevent the model from overfitting to early samples and empirically useful when you scale to a very large batch size and learning rate. You can change warm-up params to determine how long the phase is
> 
> Thanks for your reply. However, we've also done an experiment without LR schedualer, and the loss curve was still strange as I described above. On the other hand, the loss curve of GPT model is smooth and normal all the time (w or w/o LR schedualer). Only the one of BERT is confusing.
> 
> We've double-checked all our experimental settings, and there're also other researchers except for us facing the same confusing question.

Would you provide more context? Like how many steps have you trained and what is the loss value, how does the curve look like etc.

**Comment by hobbitlzy** [comment]
I also meet the problem that the loss shows two sharp decrease. Have you understood the reason?

**Comment by lihaoxin2020** [comment]
Exact same scenario to me. My first reaction was learning rate too large, then I decrease initial lr to 2.6e-5. It seems interesting that no matter I set it 1e-4 or 2.6e-5 it always plateaus around the same loss value. Looking forward to reasonable explanations.
